### ISSUE TITLE
[HOLD for payment 2024-02-14] [$500]  Login - When try to log in with Google, an error message under email field appears

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#31413]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.0-0
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
1. Go to https://staging.new.expensify.com/
2. Click on compose box
3. Click on the Google button to log in.

## Expected Result:
No error message should be displayed

## Actual Result:
An error message under email field appears saying that you must enter an email or phone number

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] Windows: Chrome / 
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/dc2e1c72-3a00-481b-a321-ae895e7734f2

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01898b63da3ac78d6c</li>
        <li>Upwork Job ID: 1725561784156704768</li>
        <li>Last Price Increase: 2023-11-24</li>
<li>Automatic offers: </li>
<ul>
<li>aimane-chnaif | Contributor | 27870856</li>
<li>FitseTLT | Contributor | 28071207</li>
</ul></li>
    </ul>
</details>

**Comment by s-alves10** [comment]
https://github.com/Expensify/App/issues/29640 would fix this issue

**Comment by roryabraham** [comment]
Confirmed reproducible on staging but not main

**Comment by roryabraham** [comment]
Not going to treat this as a blocker, and we'll see if it's fixed on staging after the next staging deploy. I don't actually think it will be – https://github.com/Expensify/App/pull/30563 is already on staging.

**Comment by Mihai-Damaschin** [comment]
I run the project and checked the issue, the simplest way to solve it is to add new state to the LoginForm component from BaseLoginForm file. New state will be changed to true when user make click on the apple or google log in buttons, and will prevent the validate function from onBlur event of TextInput component.

I already tested this solution it.

**Comment by Fralalala** [comment]
## Proposal
- Remove the code the focuses on the input form in `BaseLoginForm.js`

### Please re-state the problem that we are trying to solve in this issue.
- Login - When try to log in with Google, an error message under email field appears

### What is the root cause of that problem?
- The validation function `validate` is meant to run whenever you unfocus from the email input field. This means that since we are initially focused on the input field and we try to click anything else besides the input field (in this case the google log in button), we will trigger the `validate` function from the `onBlur` property of the component.

### What changes do you think we should make in order to solve the problem?
- We remove the code that automatically focuses the input form. 

We have 2 things in the code base:
1. Initial focus on the email input field.
2. `validate` function runs whenever you unfocus from the input field.

Number 1 and Number 2 assumes that the user will write an email at the start. If they dont intend to, say they want to use google log in or even just click or tap any thing outside of the input field, it will validate and give an error since they didnt input anything.

### What alternative solutions did you explore? (Optional)
Alternative 1: validate on input change
- For the `TextInput` component in the `BaseLoginForm.js`, we remove the `onBlur` prop. 
- After removing that prop, inside of the `onTextInput` function, remove the if condition regarding using the `validate` function. 
- After removing the if statement, it will now validate whenever the user input something in the email input field. No errors will show by simply blurring from the input field.

**Comment by Fralalala** [comment]
Contributor details
Your Expensify account email: rolando.moralesjr21@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01761a37d350487b73

**Comment by roryabraham** [comment]
Few thoughts:

- I am suspicious of the `firstBlurred` variable:
   - It is a boolean so should be prefixed by `is` or `should`
   - It seems flipped? We prevent validation on blur only if `firstBlurred` is `true`, which seems like the opposite of what we want.
- I like the idea of validating on input change rather than blur. Seems simple and expected.

@aimane-chnaif @robertKozik @Beamanator @luacmartins what do you think?

**Comment by luacmartins** [comment]
According to our [Form guidelines](https://github.com/Expensify/App/blob/c6f922947f3c3ea283130d21addd15ed8ac660ec/contributingGuides/FORMS.md#validate-on-blur-on-change-and-submit), we do validate on change, but only after the initial blur. Are we proposing that we change that?

**Comment by roryabraham** [comment]
Ok, so maybe we should update this so that when the input is blurred, we only validate if it the value was ever changed? That way if you open a page, and it auto-focuses an input, and you click away, we don't show an error.

Alternatively, we could just close this?

**Comment by peterdbarkerUK** [comment]
Not overdue, US folks are OoO

**Comment by Fralalala** [comment]
@aimane-chnaif  My bad must've interchanged staging with dev while testing. 

Retested, bug is not reproducible on dev. The `e.preventDefault()` code fixed the bug.

**Comment by roryabraham** [comment]
this one is tricky. The code seems to work on dev. During my initial investigation I checked out the staging branch, ran the app locally, and found that it was not reproducible in the dev environment. So there must be some subtle difference between the dev bundle and the staging bundle to explain this bug. Maybe even a bug in babel or something?

**Comment by roryabraham** [comment]
[discussing internally](https://expensify.slack.com/archives/C01SKUP7QR0/p1701982459692609)

**Comment by roryabraham** [comment]
Top notch sleuthing @aimane-chnaif!!!

**Comment by roryabraham** [comment]
Brought this up in slack [here](https://expensify.slack.com/archives/C04KQL2HSQ1/p1702489818284219)

**Comment by roryabraham** [comment]
@lindboe from IR suggests:

> I'm not sure about anything we can do about that. We've configured the Google library the way it requires us to, and we can't alter how it's rendering the iframe.
I think the fact that clicking elsewhere on the page also triggers the issue indicates that the existing pattern doesn't suit this form

**Comment by peterdbarkerUK** [comment]
DM'd Rory a couple questions here

**Comment by roryabraham** [comment]
> So the root cause is that preventDefault doesn't work for iframe

Currently I'm trying to dissect what `preventDefault` we're talking about. My understanding is that:

- we've got [this View with `onMouseDown` preventing default](https://github.com/Expensify/App/blob/a3c4288a20f3c77e157cbe3ab0c0e26a8dbbe8e5/src/pages/signin/LoginForm/BaseLoginForm.js#L319-L321), wrapping the Google Sign In button.
- On dev, that `preventDefault` prevents the focus from leaving the `TextInput`, so the form validation doesn't happen
- However, on prod, an iframe is rendered and, before it's even rendering anything visible, captures focus. This blurs the `TextInput` and causes it to validate prematurely.

Still not sure how best to solve this, but a couple ideas to address this could be:

- Adjust our form guidelines such that `onBlur` only triggers a validation if the focus moves to some other element in the same form.
    - We could do this by traversing the DOM upwards from the `document.activeElement` looking for the form root or downwards from the form root looking for the `document.activeElement`.
- Adjust our form guidelines such that `onBlur` only triggers a validation if the focus moves to some other element in the same window.
    - We could do this by simply checking `document.hasFocus`

**Comment by roryabraham** [comment]
Posted discussion in slack: https://expensify.slack.com/archives/C01GTK53T8Q/p1703701749775979

**Comment by roryabraham** [comment]
Seems we hav consensus on this solution:

> Adjust our form guidelines such that onBlur only triggers a validation if the focus moves to some other element in the same HTML document. We can do this by using `document.hasFocus`

Going to make this external again and will review proposals to implement this change

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When try to log in with Google, an error message under email field appears

### What is the root cause of that problem?
This is because we are validating although the document is not in focus here
https://github.com/Expensify/App/blob/7d1b0f8b99616c48aedb975dd73fc3086c9f20e3/src/pages/signin/LoginForm/BaseLoginForm.js#L275-L276
### What changes do you think we should make in order to solve the problem?

we should change it to 
```
if (!document.hasFocus) return;
                        validate(login);
```
### What alternative solutions did you explore? (Optional)

**Comment by roryabraham** [comment]
Hey, @FitseTLT when can we expect a PR to be ready for review?

**Comment by roryabraham** [comment]
> we can call the validate code inside runAfterInteraction inside onBlur of the text input but the downside would be when we blur the input by clicking other than the google signin the validation error will appear with delay

I like this idea better. By "appear with delay", do you mean only after the few milliseconds of delay caused by `runAfterInteractions`? If yes, I think that's acceptable. However, I think on web there might be a problem with `runAfterInteractions` in that lottie animations can cause it to never be triggered. Not sure if that's still the case of if that would be a problem here.

**Comment by FitseTLT** [comment]
Here is how it looks. For me it is noticeable, I remember a case where a delay caused by `runAfterInteraction` was reported as a regression but let me hear your thoughts. And don't forget it also depends on device performance.

https://github.com/Expensify/App/assets/38649957/2443d244-7c47-45fd-a5ee-d6428bbac4e4

**Comment by roryabraham** [comment]
I think that delay looks ok, but I think it would be less if the lottie animation wasn't interfering with `runAfterInteractions`. I think we can treat that as out-of-scope for this issue, but it keeps coming up. We should probably create a follow-up issue to try and find a solution to that problem

**Comment by roryabraham** [comment]
To expand on that, on web `InteractionsManager.runAfterInteractions` uses [window.requestIdleCallback](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestIdleCallback), or a [50ms `setTimeout` polyfill]( https://github.com/necolas/react-native-web/blob/5f9c32eba0f840bde7b462a57d0748358ff866f6/packages/react-native-web/src/modules/requestIdleCallback/index.js)

**Comment by roryabraham** [comment]
So maybe we need to get a bit hacky here with something like this?

```js
onBlur={() => {
    Promise.race([new Promise((resolve) => InteractionManager.runAfterInteractions(resolve)), new Promise((resolve) => setTimeout(resolve, 50))]).then(() => {
        if (firstBlurred.current || !Visibility.isVisible() || !Visibility.hasFocus()) {
            return;
        }
        firstBlurred.current = true;
        validate(login);
    });
}}
```

Maybe we can build an abstraction around that like `runAfterInteractionsOrMaxTimeToWait`

Honestly not sure if this is a good idea or not. What do you think?

**Comment by FitseTLT** [comment]
I don't think that will help on the delay. Maybe shall I create a PR with my 2nd approach??
>We can call the validate code when clicking on the sign in page and the textInput is not focused. We can easily achieve this by imperative handle in login form to be called in Signin Page.

I mean I know we should stick to declarative approach as much as possible but there is a limit to that and I think this could be a perfect situation we should utilize imerative code.

**Comment by roryabraham** [comment]
> I mean I know we should stick to declarative approach as much as possible but there is a limit to that and I think this could be a perfect situation we should utilize imerative code.

My concern with this approach is that I don't know how it will translate consistently to other forms in different places throughout the codebase. It's a pattern that would not be easy to employ consistently.

**Comment by roryabraham** [comment]
> I don't think that will help on the delay

Why not? Won't it mean that the delay will be at most 50ms?

**Comment by FitseTLT** [comment]
>My concern with this approach is that I don't know how it will translate consistently to other forms in different places throughout the codebase. It's a pattern that would not be easy to employ consistently.

As I mentioned before this is an exceptional issue related with only `BaseLoginForm`. I am not planning to apply the solution for other forms at all.
>Why not? Won't it mean that the delay will be at most 50ms?
I mean we are in a trade of here if we wait until the end of the interaction we are solving the issue but introducing lag or else we will not solve the issue if we limit the delay and the interaction takes longer.

**Comment by roryabraham** [comment]
> I am not planning to apply the solution for other forms at all.

Exactly my concern.

> I mean we are in a trade of here if we wait until the end of the interaction we are solving the issue but introducing lag

Right, but normally the interaction would only last a few ms and not be perceptible by the human eye – it's just that it's delayed to last hundreds or thousands of ms due to the lottie animation making `window.requestIdleCallback` take longer. So by capping that to be at most 50ms, we ensure that it will not be perceptible by the human eye (rule of thumb says that anything less than 250ms will not be noticeable)

**Comment by FitseTLT** [comment]
>
I am not planning to apply the solution for other forms at all.

Exactly my concern.

I mean we are in a trade of here if we wait until the end of the interaction we are solving the issue but introducing lag

Right, but normally the interaction would only last a few ms and not be perceptible by the human eye – it's just that it's delayed to last hundreds or thousands of ms due to the lottie animation making window.requestIdleCallback take longer. So by capping that to be at most 50ms, we ensure that it will not be perceptible by the human eye (rule of thumb says that anything less than 250ms will not be noticeable)

So no need for `Promise.race` we just decide some delay value and proceed hoping it will not be shorter than the new page opening delay ??

**Comment by roryabraham** [comment]
> So no need for Promise.race we just decide some delay value and proceed hoping it will not be shorter than the new page opening delay ??

Right, good point – we need to time the `validate` to happen _after_ the Google Sign In iframe has loaded, but _before_ it becomes visible

**Comment by FitseTLT** [comment]
@roryabraham On my PC 300 ms is the edge case when I decrease it to 250 ms the error message appears and 300ms works well. I mean the `runAfterInteraction` takes a lot of time so if we chose the delay approach What delay should I chose? Let me know that and will create PR tomorrow 👍

**Comment by roryabraham** [comment]
I'm not sure – if 300ms works for you but 250ms does not, it sounds like that's cutting it a little close. So let's bump it up to 500ms and see how that works?

**Comment by roryabraham** [comment]
Long-term we really need to try to find a way to make it so lottie doesn't with `runAfterInteractions` / `requestIdleCallback`

**Comment by roryabraham** [comment]
Ok, just to clarify next steps, I think @FitseTLT will open a PR to fix this using `setTimeout` as discussed above. @FitseTLT let us know when we can expect a PR

**Comment by FitseTLT** [comment]
Definitely in this weekend! Sorry, I was busy, didn't forget it 👍

**Comment by roryabraham** [comment]
@anmurali this is ready for payment.

**Comment by anmurali** [comment]
@FitseTLT has been paid
@aimane-chnaif offer expired, so sent a new one [here](https://www.upwork.com/nx/wm/offer/100942703)
Do we need to add a regression test? If yes, @aimane-chnaif can you propose it?

Adding a backup BZ member to close this off after those two things are handled here as I am OOO

**Comment by adelekennedy** [comment]
@aimane-chnaif paid! Bump on the regression test question

**Comment by adelekennedy** [comment]
perfect - then we can close this!
